### PR TITLE
8304725: AsyncGetCallTrace can cause SIGBUS on M1

### PR DESCRIPTION
**Pull request by parttimenerd**
Fixes the issue by disabling PCDesc cache modifications when in ASGCT.

Tested on my M1 mac.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8304725](https://bugs.openjdk.org/browse/JDK-8304725): AsyncGetCallTrace can cause SIGBUS on M1

### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**) ⚠️ Review applies to [6f1108ed](https://git.openjdk.org/jdk/pull/13144/files/6f1108ed4a4ffea3f8f784140834b0df3f21daa5)
 * [Thomas Stuefe](https://openjdk.org/census#stuefe) (@tstuefe - **Reviewer**) ⚠️ Review applies to [468d71e3](https://git.openjdk.org/jdk/pull/13144/files/468d71e370e9ba6334022b46ceae0680b7686d64)
 * [Matthias Baesken](https://openjdk.org/census#mbaesken) (@MBaesken - **Reviewer**) ⚠️ Review applies to [6f1108ed](https://git.openjdk.org/jdk/pull/13144/files/6f1108ed4a4ffea3f8f784140834b0df3f21daa5)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/13144/head:pull/13144` \
`$ git checkout pull/13144`

Update a local copy of the PR: \
`$ git checkout pull/13144` \
`$ git pull https://git.openjdk.org/jdk.git pull/13144/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 13144`

View PR using the GUI difftool: \
`$ git pr show -t 13144`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/13144.diff">https://git.openjdk.org/jdk/pull/13144.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/13144#issuecomment-1479878306)

### CODE REVIEW

**Review by dholmes-ora — APPROVED** [review]
Seems okay. But I am really over this game of whack-a-mole with ThreadWXEnable.

**Comment by parttimenerd** [comment]
> Seems okay. But I am really over this game of whack-a-mole with ThreadWXEnable.

You're not the only one. 

Andrei Pangin added a workaround to async-profiler, so nobody noticed the problem before.

**Comment by tstuefe** [comment]
@parttimenerd I think you need two reviews in hotspot for integration.

You now effectively disable execution of generated code for the whole extend of AGCT? So we cannot call stub routines anymore. Won't hurt safefetch since it uses static assembly now, but something to keep in mind if you ever want to downport this.

**Review by tstuefe — APPROVED** [review]
Okay. Though the prospect of async profiler modifying the code cache by walking the stack seems scary.

**Comment by fisk** [comment]
Drive by comment: how async safe is WX enabler? If a thread is in the middle of it and we shoot a signal and enable, what will happen?

**Comment by parttimenerd** [comment]
> You now effectively disable execution of generated code for the whole extend of AGCT?

That's exactly what async-profiler does already https://github.com/async-profiler/async-profiler/blob/c8de91df6b090af82e91a066deb81a3afb505331/src/profiler.cpp#L383, I wonder why we don't have problems there with SafeFetch on older JVMs.

> Okay. Though the prospect of async profiler modifying the code cache by walking the stack seems scary.

We discussed it before. It's probably safe, but yes, my initial reaction was also "why not just remove this", but it should have a performance impact.

> Drive by comment: how async safe is WX enabler? If a thread is in the middle of it and we shoot a signal and enable, what will happen?

This is a really good point. I think that making the field that stores this information `volatile` could alleviate the problem (?). This would ensure that no reordering takes place in:

```
 _wx_state = WXWrite;
  os::current_thread_enable_wx(_wx_state);
```
(https://github.com/openjdk/jdk/blob/77cd917a97b184871ab2d3325ceb6c53afeca28b/src/hotspot/share/runtime/thread.inline.hpp#L78)

**Comment by fisk** [comment]
> > You now effectively disable execution of generated code for the whole extend of AGCT?
> 
> 
> 
> That's exactly what async-profiler does already https://github.com/async-profiler/async-profiler/blob/c8de91df6b090af82e91a066deb81a3afb505331/src/profiler.cpp#L383, I wonder why we don't have problems there with SafeFetch on older JVMs.
> 
> 
> 
> > Okay. Though the prospect of async profiler modifying the code cache by walking the stack seems scary.
> 
> 
> 
> We discussed it before. It's probably safe, but yes, my initial reaction was also "why not just remove this", but it should have a performance impact.
> 
> 
> 
> > Drive by comment: how async safe is WX enabler? If a thread is in the middle of it and we shoot a signal and enable, what will happen?
> 
> 
> 
> This is a really good point. I think that making the field that stores this information `volatile` could alleviate the problem (?). This would ensure that no reordering takes place in:
> 
> 
> 
> ```
> 
>  _wx_state = WXWrite;
> 
>   os::current_thread_enable_wx(_wx_state);
> 
> ```
> 
> (https://github.com/openjdk/jdk/blob/77cd917a97b184871ab2d3325ceb6c53afeca28b/src/hotspot/share/runtime/thread.inline.hpp#L78)

If you look at the enable_wx function, there is a lack of atomicity of the operation. It checks the current state and only conditionally decides to change the state. And when it does there is a write of the new state and a call to actually flip the mode. Seems to me that there could be many problematic interleavings where the signal hits in this code, which could mess things up. This code was not designed to be async safe, and I'm not convinced that sprinkling in volatile really solves that problem.

**Comment by parttimenerd** [comment]
The best alternative to me is to take the perf hit and disable the caching when we're in forte (possibly only on Mac).

**Comment by fisk** [comment]
> The best alternative to me is to take the perf hit and disable the caching when we're in forte (possibly only on Mac).

Sounds like a plan.

**Comment by parttimenerd** [comment]
We could just add a method `is_writable()` at https://github.com/openjdk/jdk/blob/8af25ec09d47900998b8fb58e2e7d486420df03d/src/hotspot/share/runtime/thread.hpp#L629 and call it in PcDesc code.

**Comment by tstuefe** [comment]
Reading @fisk excellent catch the async safety of stacking the wx raii mechanics, I retract my approval. This looks like a recipe for hard-to-find bugs :-/

Incidentally, do we see in the hs_err file whether async profiler is attached? We should maybe make that prominently visible.

> > The best alternative to me is to take the perf hit and disable the caching when we're in forte (possibly only on Mac).
> 
> Sounds like a plan.

I feel there are information missing :)

**Comment by parttimenerd** [comment]
> Reading @fisk excellent catch the async safety of stacking the wx raii mechanics, I retract my approval. This looks like a recipe for hard-to-find bugs :-/

Yes, this is my current thought too.

> Incidentally, do we see in the hs_err file whether async profiler is attached? We should maybe make that prominently visible.

We see ASGCT in the stack trace, but what exactly do you mean.

> I feel there are information missing :)

But would you be okay with disabling the cache modification if the perf impact is low?

**Comment by tstuefe** [comment]
> > Reading @fisk excellent catch the async safety of stacking the wx raii mechanics, I retract my approval. This looks like a recipe for hard-to-find bugs :-/
> 
> Yes, this is my current thought too.
> 
> > Incidentally, do we see in the hs_err file whether async profiler is attached? We should maybe make that prominently visible.
> 
> We see ASGCT in the stack trace, but what exactly do you mean.
> 

For issues like this, you wouldn't necessarily have AGCT on the stack. Consider:
- Compiler gets invoked, write protects, compiles, then tries to restore write protection. Gets interrupted by AGCT after calling pthread_jit_write_protect_np but before updating Thread::_wx_state.
- Now AGCT runs. It disables write protection, does its thing, then reinstates the state *it thinks preceded it*. But that is the wrong state. In this case, the one used during compilation.
- we return from signal handling. Compiler now sets Thread::_wx_state and assumes write protection is restored, but it isnt.
- Later that day, the thread tries to call into compiled code. It will not be able to execute it, since the protection is wrong. 

There are variants of this play, but my point is the resulting crashes may happen after AGCT was invoked. 

So all we spy with our little eyes would be a segfault, I guess SEGV_ACCERR ?, and maybe the AGCT shared lib among the list of loaded libraries. 

In particular, we do not know if AGCT did interrupt the crashing thread recently. Or do we? This would be valuable information.

**Review by dholmes-ora — CHANGES_REQUESTED** [review]
So AGCT already breaks in this scenario because WX is not enabled when needed. The code to enable WX is not async-signal_safe, so the "fix" is only a partial fix - it will sometimes fail too. So we go from broken to slightly less broken.

This is why AGCT is not supported.

Anyway retracting my review while discussion continues.

**Comment by dholmes-ora** [comment]
Correction: I missed some of the discussion above. I see now this AGCT "fix" can break other code, so this makes things much more broken.

**Comment by tstuefe** [comment]
> > In particular, we do not know if AGCT did interrupt the crashing thread recently. Or do we? This would be valuable information.
> 
> Yes. But it could indeed be helpful information when debugging problems. Please don't forget that JFR has a very similar code, and it would be good to have the information on whether JFR sampled a thread recently too.

Yes, but while JFR interrupts threads too, its sampler runs in its own thread, so the async-safety of the interrupted code should not matter, or?

wrt info, just a little marker in the thread "AGCT was here", maybe with a timestamp, would be useful.

**Comment by parttimenerd** [comment]
> Yes, but while JFR interrupts threads too, its sampler runs in its own thread, so the async-safety of the interrupted code should not matter, or?

You're right.

> wrt info, just a little marker in the thread "AGCT was here", maybe with a timestamp, would be useful.

That sounds like a great idea.

**Comment by jbachorik** [comment]
> Yes, but while JFR interrupts threads too, its sampler runs in its own thread, so the async-safety of the interrupted code should not matter, or?

And this sampling approach is quite subpar when compared with the signal based profiling - both in terms of bias and precision. Meaning that if/when JFR would implement a state-of-the-art CPU profiler it will have to deal with the same issues we are seeing here.

**Comment by tstuefe** [comment]
> > Yes, but while JFR interrupts threads too, its sampler runs in its own thread, so the async-safety of the interrupted code should not matter, or?
> 
> And this sampling approach is quite subpar when compared with the signal based profiling - both in terms of bias and precision. Meaning that if/when JFR would implement a state-of-the-art CPU profiler it will have to deal with the same issues we are seeing here.

Not if they run the walker outside the sampled thread. Which would be much safer in a lot of ways.

**Comment by parttimenerd** [comment]
> Not if they run the walker outside the sampled thread. Which would be much safer in a lot of ways.

Yes, it would, but it is not currently possible with ASGCT, albeit changing it would be quite simple (not changing existing use cases or the API, but would require more tests). I would rather add this to the new JEP.

File: src/hotspot/share/runtime/thread.hpp [hunk 1]
```diff
@@ -631,6 +631,31 @@ class Thread: public ThreadShadow {
...
   }
 #endif // __APPLE__ && AARCH64
+
+ private:
+  bool _in_asgct = false;
+ public:
+  bool in_asgct() const { return _in_asgct; }
+  void set_in_asgct(bool value) { _in_asgct = value; }
+  static bool current_in_asgct() {
+    Thread *cur = Thread::current();
```
Comment (dholmes-ora) [inline src/hotspot/share/runtime/thread.hpp:641]: You need to use`current_or_null_safe` here as you may be in a signal handling context.

File: src/hotspot/share/runtime/thread.hpp [hunk 2]
```diff
@@ -631,6 +631,31 @@ class Thread: public ThreadShadow {
...
+    return cur != nullptr && cur->in_asgct();
+  }
+};
+
+class ThreadInAsgct {
+ private:
+  Thread* _thread;
+ public:
+  ThreadInAsgct(Thread* thread) : _thread(thread) {
+    assert(thread != NULL, "invariant");
```
Comment (dholmes-ora) [inline src/hotspot/share/runtime/thread.hpp:651]: s/NULL/nullptr/

**Review by dholmes-ora — CHANGES_REQUESTED** [review]

**Comment by parttimenerd** [comment]
Thanks for the review, I fixed the issues.

**Comment by dholmes-ora** [comment]
Can you update the description please to reflect the final approach.

**Comment by dholmes-ora** [comment]
Can I just clarify the new approach please. IIUC if we update the PcDesc cache whilst within ASGCT then we hit code that needs WXWrite mode. So the fix is to just skip updating the cache if ASGCT is active. I have no idea what the PcDesc cache is but I presume if it is really a cache then not updating it only potentially affects lookup performance - right?

Thanks.

**Comment by parttimenerd** [comment]
You're right and I updated the description of the PR. I gave some performance figures in a previous comment, they show that the performance impact is probably negligible.

**Review by dholmes-ora — APPROVED** [review]
Nothing further from me. Thanks.

**Review by MBaesken — APPROVED** [review]
LGTM.
I would like a short comment line in nmethod.cpp and/or forte.cpp shortly describing what you do and why but up to you.

**Comment by parttimenerd** [comment]
I added two comments :)

**Pull request closed without merging**